src/react-integration/provider/__tests__/middleware.tsx

```diff
@@ -139,6 +152,39 @@ describe('createEnhancedReducerHook', () => {
     expect(faker.mock.calls[2][0]).toEqual({ type: 'nothing', payload: 5 });
   });
 
+  test('should work with middlewares that getState()', () => {
+    const callBefore = jest.fn();
+    const callAfter = jest.fn();
+    const logger = makeStatefulMiddleware({ callBefore, callAfter });
+
+    const reducer = (state: { counter: number }) => ({
+      ...state,
+      counter: state.counter + 1,
+    });
+    const { result } = renderHook(() => {
+      const useEnhancedReducer = createEnhancedReducerHook(logger);
+      return useEnhancedReducer(reducer, { counter: 0 });
+    });
+    let [state, dispatch] = result.current;
+    expect(callBefore.mock.calls.length).toBe(0);
+    let action: any = { type: 'hi' };
+    act(() => {
+      dispatch(action);
+    });
+    [state, dispatch] = result.current;
+    expect(callBefore.mock.calls.length).toBe(1);
+    //expect(callAfter.mock.calls.length).toBe(1);
```

Comment:
i think this doesn't work because for store to be updated it not only has to finish the reducer, but actually run render in the react function.

line 16 of middleware.tsx:
```tsx
store.current.state = state;
```

I wonder if there's a mechanism to block the rest of dispatch until that commit takes place?

Reply:
Yes that's where `useReducer` behaves not exactly like Redux. One way to solve this is to manually call the reducer and store the state in a ref: https://github.com/sylvaingi/rest-hooks/commit/c1ad7e3fd810e97790bb2c942c95b15970288127

Reply:
Hmm, actually then if you just use scheduler (like in https://github.com/coinbase/rest-hooks/pull/64) you should be able to get the next value after it commits.

Reply:
And persisting should honestly be done async on a scheduler anyway - with a lower priority even. So probably works out fine.